lib/src/zap/sock/zap_sock.c

```diff
@@ -256,12 +256,12 @@ static zap_err_t z_sock_close(zap_ep_t ep)
 	case ZAP_EP_CONNECTED:
 	case ZAP_EP_LISTENING:
 		sep->ep.state = ZAP_EP_CLOSE;
-		// shutdown(sep->sock, SHUT_RDWR);
+		shutdown(sep->sock, SHUT_RDWR);
 		break;
 	case ZAP_EP_ERROR:
 	case ZAP_EP_ACCEPTING:
 	case ZAP_EP_CONNECTING:
-		// shutdown(sep->sock, SHUT_RDWR);
+		shutdown(sep->sock, SHUT_RDWR);
```

Comment:
shouldn't the endpoint state be set to ZAP_EP_CLOSE here? Seems like there is a race with sock_event where the endpoint could be left in ZAP_EP_ACCEPTING/CONNECTING, etc...

Reply:
Let me look into it .. I don't wanna break the state machine in the last minute :worried:

Reply:
@tom95858 We do not move the state to ZAP_EP_CLOSE because at this point, the underlying socket already became `connected` and we expect `EPOLLHUP` from epoll. The EPOLLHUP will be processed in `sock_event()` function. If we move the endpoint state to `ZAP_EP_CLOSE`, `sock_event()` will incorrectly deliver `ZAP_EP_DISCONNECTED` event, and that breaks zap event design (DISCONNECTED must only be delivered if CONNECTED has been delivered). By not moving the state in these cases, `sock_event()` knows that the underlying socket terminated while the zap endpoint is in connecting/accepting states and then delivers corresponding CONNECT_ERROR events.